dom.go

```diff
@@ -1,92 +1,164 @@
 package vecty
 
-import "github.com/gopherjs/gopherjs/js"
+import (
+	"fmt"
 
-// Component represents a Vecty component.
-type Component interface {
-	Markup
-	Reconcile(oldComp Component)
-	Node() *js.Object
-}
+	"github.com/gopherjs/gopherjs/js"
+)
 
-// Render renders a component into the given container element. It is appended
-// as a child element.
-func Render(comp Component, container *js.Object) {
-	comp.Reconcile(nil)
-	container.Call("appendChild", comp.Node())
+// Core implements the Context method of the Component interface, and is the
+// core/central struct which all Component implementations should embed.
+type Core struct {
+	prevRender *HTML
 }
 
-// RenderAsBody renders the given component as the body of the page, replacing
-// whatever existing content in the page body there may be.
-func RenderAsBody(comp Component) {
-	doc := js.Global.Get("document")
-	body := doc.Call("createElement", "body")
-	Render(comp, body)
-	if doc.Get("readyState").String() == "loading" {
-		doc.Call("addEventListener", "DOMContentLoaded", func() { // avoid duplicate body
-			doc.Set("body", body)
-		})
-		return
-	}
-	doc.Set("body", body)
-}
+// Context implements the Component interface.
+func (c *Core) Context() *Core { return c }
 
-type textComponent struct {
-	text string
-	node *js.Object
-}
+// Component represents a single visual component within an application. To
+// define a new component simply implement the Render method and embed the Core
+// struct:
+//
+// 	type MyComponent struct {
+// 		vecty.Core
+// 		... additional component fields (state or properties) ...
+// 	}
+//
+// 	func (c *MyComponent) Render() *vecty.HTML {
+// 		... rendering ...
+// 	}
+//
```

Comment:
Component creation is greatly simplified (and documented) with this new pattern. `Apply` is no longer needed for components (which was confusing from a user perspective), and `Restore` is optional / a default is provided. It means you only need to use what you want.